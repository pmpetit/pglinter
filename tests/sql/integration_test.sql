-- Comprehensive integration test for pglinter
-- Tests multiple rules across all categories (B, C, T, S series)
BEGIN;

DROP EXTENSION IF EXISTS pglinter CASCADE;

-- Create various test scenarios to trigger multiple rules

-- 1. Tables without primary keys (B001, T001)
CREATE TABLE users_no_pk (
    id INT,
    username TEXT,
    email TEXT
);

-- 2. Redundant indexes (B002, T003)
CREATE TABLE products (
    id SERIAL PRIMARY KEY,
    name TEXT,
    category TEXT,
    price NUMERIC
);

CREATE INDEX idx_name_1 ON products (name);
CREATE INDEX idx_name_2 ON products (name); -- redundant
CREATE INDEX idx_composite_1 ON products (category, price);
CREATE INDEX idx_composite_2 ON products (category, price); -- redundant

-- 3. Foreign keys without indexes (B003, T004)
CREATE TABLE orders (
    id SERIAL PRIMARY KEY,
    user_id INT,
    product_id INT,
    total NUMERIC
);

-- Add foreign keys without creating indexes
ALTER TABLE orders ADD CONSTRAINT fk_user
FOREIGN KEY (user_id) REFERENCES users_no_pk (id);
ALTER TABLE orders ADD CONSTRAINT fk_product
FOREIGN KEY (product_id) REFERENCES products (id);

-- 4. Tables with uppercase names/columns (B006, T011)
CREATE TABLE "UPPERCASE_ISSUES" (
    id SERIAL PRIMARY KEY,
    "UPPER_NAME" TEXT,
    "DESCRIPTION" TEXT
);

-- 5. Reserved keywords (T010)
CREATE TABLE items (
    id SERIAL PRIMARY KEY,
    "SELECT" TEXT, -- reserved keyword
    "FROM" TEXT,   -- reserved keyword
    description TEXT
);

-- 6. Environment-prefixed schemas (S002)
CREATE SCHEMA prod_analytics;
CREATE SCHEMA dev_reports;
CREATE SCHEMA staging_data;

-- Add some tables to the schemas
CREATE TABLE prod_analytics.metrics (
    id SERIAL PRIMARY KEY,
    metric_name TEXT
);

CREATE TABLE dev_reports.summaries (
    id SERIAL PRIMARY KEY,
    report_data TEXT
);

-- Create extension
CREATE EXTENSION IF NOT EXISTS pglinter;

-- Run comprehensive analysis
SELECT '=== COMPREHENSIVE pglinter ANALYSIS ===' AS info;

-- Test all rule categories
SELECT 'BASE RULES:' AS category;
SELECT pglinter.perform_base_check();

SELECT 'TABLE RULES:' AS category;
SELECT pglinter.perform_table_check();

SELECT 'SCHEMA RULES:' AS category;
SELECT pglinter.perform_schema_check();

SELECT 'CLUSTER RULES:' AS category;
SELECT pglinter.perform_cluster_check();

-- Test comprehensive check
SELECT 'COMPREHENSIVE CHECK:' AS category;
SELECT pglinter.check_all();

-- Test rule management features
SELECT '=== RULE MANAGEMENT ===' AS info;

-- Show all rules
SELECT pglinter.show_rules();

-- Test some explanations
SELECT pglinter.explain_rule('B001');
SELECT pglinter.explain_rule('T003');
SELECT pglinter.explain_rule('S002');

-- Test output to file functionality
SELECT '=== OUTPUT TO FILE TEST ===' AS info;
SELECT pglinter.perform_base_check('/tmp/integration_base.sarif');
SELECT pglinter.perform_table_check('/tmp/integration_table.sarif');
SELECT pglinter.perform_schema_check('/tmp/integration_schema.sarif');

-- Test rule disable/enable functionality
SELECT '=== RULE TOGGLE TEST ===' AS info;

-- Disable some rules
SELECT pglinter.disable_rule('B001');
SELECT pglinter.disable_rule('T003');

-- Run checks (should skip disabled rules)
SELECT pglinter.perform_base_check();
SELECT pglinter.perform_table_check();

-- Re-enable rules
SELECT pglinter.enable_rule('B001');
SELECT pglinter.enable_rule('T003');

-- Final comprehensive check
SELECT pglinter.check_all();

-- Clean up
DROP SCHEMA prod_analytics CASCADE;
DROP SCHEMA dev_reports CASCADE;
DROP SCHEMA staging_data CASCADE;
DROP TABLE items CASCADE;
DROP TABLE "UPPERCASE_ISSUES" CASCADE;
DROP TABLE orders CASCADE;
DROP TABLE products CASCADE;
DROP TABLE users_no_pk CASCADE;

ROLLBACK;
