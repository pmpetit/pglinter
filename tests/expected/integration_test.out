-- Comprehensive integration test for pglinter
-- Tests multiple rules across all categories (B, C, T, S series)
CREATE EXTENSION pglinter;
-- Create various test scenarios to trigger multiple rules
-- 1. Tables without primary keys (B001, T001)
CREATE TABLE users_no_pk (
    id INT,
    username TEXT,
    email TEXT
);
-- 2. Redundant indexes (B002, T003)
CREATE TABLE products (
    id SERIAL PRIMARY KEY,
    name TEXT,
    category TEXT,
    price NUMERIC
);
CREATE INDEX idx_name_1 ON products (name);
CREATE INDEX idx_name_2 ON products (name); -- redundant
CREATE INDEX idx_composite_1 ON products (category, price);
CREATE INDEX idx_composite_2 ON products (category, price); -- redundant
-- 3. Foreign keys without indexes (B003, T004)
CREATE TABLE orders (
    id SERIAL PRIMARY KEY,
    user_id INT,
    product_id INT,
    total NUMERIC
);
-- Add foreign keys without creating indexes
ALTER TABLE orders ADD CONSTRAINT fk_user
FOREIGN KEY (user_id) REFERENCES users_no_pk (id);
ERROR:  there is no unique constraint matching given keys for referenced table "users_no_pk"
ALTER TABLE orders ADD CONSTRAINT fk_product
FOREIGN KEY (product_id) REFERENCES products (id);
-- 4. Tables with uppercase names/columns (B006, T011)
CREATE TABLE "UPPERCASE_ISSUES" (
    id SERIAL PRIMARY KEY,
    "UPPER_NAME" TEXT,
    "DESCRIPTION" TEXT
);
-- 5. Reserved keywords (T010)
CREATE TABLE items (
    id SERIAL PRIMARY KEY,
    "SELECT" TEXT, -- reserved keyword
    "FROM" TEXT,   -- reserved keyword
    description TEXT
);
-- 6. Environment-prefixed schemas (S002)
CREATE SCHEMA prod_analytics;
CREATE SCHEMA dev_reports;
CREATE SCHEMA staging_data;
-- Add some tables to the schemas
CREATE TABLE prod_analytics.metrics (
    id SERIAL PRIMARY KEY,
    metric_name TEXT
);
CREATE TABLE dev_reports.summaries (
    id SERIAL PRIMARY KEY,
    report_data TEXT
);
-- Create extension
-- Run comprehensive analysis
SELECT '=== COMPREHENSIVE pglinter ANALYSIS ===' AS info;
                  info                   
-----------------------------------------
 === COMPREHENSIVE pglinter ANALYSIS ===
(1 row)

-- Test all rule categories
SELECT 'BASE RULES:' AS category;
  category   
-------------
 BASE RULES:
(1 row)

SELECT pglinter.perform_base_check();
NOTICE:  üîç pglinter found 3 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ùå [B001] ERROR: 16/18 table(s) without primary key exceed the error threshold: 88%.
NOTICE:  ‚ùå [B003] ERROR: 3/3 table(s) without index on foreign key exceed the error threshold: 100%.
NOTICE:  ‚ùå [B004] ERROR: 4/4 unused index exceed the error threshold: 100%
NOTICE:  ==================================================
NOTICE:  üìä Summary: 3 error(s), 0 warning(s), 0 info
NOTICE:  üî¥ Critical issues found - please review and fix errors
 perform_base_check 
--------------------
 t
(1 row)

SELECT 'TABLE RULES:' AS category;
   category   
--------------
 TABLE RULES:
(1 row)

SELECT pglinter.perform_table_check();
NOTICE:  üîç pglinter found 6 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [T001] WARNING: TABLE no primary key on table(s) 1 : 
public.users_no_pk 

NOTICE:  ‚ö†Ô∏è  [T002] WARNING: TABLE duplicated index 4 : 
public.products idx idx_composite_1 columns (category,price) is a subset of idx idx_composite_2 columns (category,price)
public.products idx idx_composite_2 columns (category,price) is a subset of idx idx_composite_1 columns (category,price)
public.products idx idx_name_1 columns (name) is a subset of idx idx_name_2 columns (name)
public.products idx idx_name_2 columns (name) is a subset of idx idx_name_1 columns (name) 

NOTICE:  ‚ö†Ô∏è  [T003] WARNING: TABLE unindexed constraint 3 : 
inventory_schema.stock_movements constraint name:stock_movements_stock_id_fkey
public.orders constraint name:fk_product
sales_schema.order_items constraint name:order_items_order_id_fkey 

NOTICE:  ‚ö†Ô∏è  [T004] WARNING: TABLE table with potential missing index 18 : 
audit_schema.user_activity_log:100 % of seq scan
dev_reports.summaries:100 % of seq scan
inventory_schema.product_stock:0 % of seq scan
inventory_schema.stock_movements:100 % of seq scan
prod_analytics.metrics:100 % of seq scan
public.UPPERCASE_ISSUES:100 % of seq scan
public.customer_orders:100 % of seq scan
public.good_table_with_pk:100 % of seq scan
public.inventory_logs:100 % of seq scan
public.items:100 % of seq scan
public.orders:100 % of seq scan
public.product_reviews:100 % of seq scan
public.products:100 % of seq scan
public.user_sessions:100 % of seq scan
public.users_no_pk:100 % of seq scan
sales_schema.order_items:100 % of seq scan
sales_schema.orders:0 % of seq scan
sales_schema.sales_reports:100 % of seq scan 

NOTICE:  ‚ö†Ô∏è  [T008] WARNING: TABLE No role grantee on table. it means that except owner, users will need a direct grant on this table, not through a role. Prefer RBAC access if possible. 8 : 
audit_schema.user_activity_log as no role granted
dev_reports.summaries as no role granted
inventory_schema.product_stock as no role granted
inventory_schema.stock_movements as no role granted
prod_analytics.metrics as no role granted
sales_schema.order_items as no role granted
sales_schema.orders as no role granted
sales_schema.sales_reports as no role granted 

NOTICE:  ‚ö†Ô∏è  [T009] WARNING: TABLE Reserved keywords in object. 2 : 
column:public.items.FROM is a reserved keyword.
column:public.items.SELECT is a reserved keyword. 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 6 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
 perform_table_check 
---------------------
 t
(1 row)

SELECT 'SCHEMA RULES:' AS category;
   category    
---------------
 SCHEMA RULES:
(1 row)

SELECT pglinter.perform_schema_check();
NOTICE:  üîç pglinter found 2 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [S001] WARNING: SCHEMA No default role grantee on schema {0}.{1}. It means that each time a table is created, you must grant it to roles. 6 : 
audit_schema
dev_reports
inventory_schema
prod_analytics
sales_schema
staging_data 

NOTICE:  ‚ö†Ô∏è  [S002] WARNING: SCHEMA You should not prefix or suffix the schema name with {0}. You may have difficulties when refreshing environments. Prefer prefix or suffix the database name. 3 : 
dev_reports
prod_analytics
staging_data 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 2 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
 perform_schema_check 
----------------------
 t
(1 row)

SELECT 'CLUSTER RULES:' AS category;
    category    
----------------
 CLUSTER RULES:
(1 row)

SELECT pglinter.perform_cluster_check();
NOTICE:  üîç pglinter found 2 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ùå [C002] ERROR: 6 entries in pg_hba.conf with trust or password authentication method exceed the warning threshold: 6.
NOTICE:  ‚ö†Ô∏è  [C003] WARNING: CLUSTER Encrypted passwords with MD5. 1 : 
password_encryption is md5 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 1 error(s), 1 warning(s), 0 info
NOTICE:  üî¥ Critical issues found - please review and fix errors
 perform_cluster_check 
-----------------------
 t
(1 row)

-- Test comprehensive check
SELECT 'COMPREHENSIVE CHECK:' AS category;
       category       
----------------------
 COMPREHENSIVE CHECK:
(1 row)

SELECT pglinter.check_all();
NOTICE:  üîç Running comprehensive pglinter check...
NOTICE:  
NOTICE:  üìã BASE CHECKS:
NOTICE:  üîç pglinter found 3 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ùå [B001] ERROR: 16/18 table(s) without primary key exceed the error threshold: 88%.
NOTICE:  ‚ùå [B003] ERROR: 3/3 table(s) without index on foreign key exceed the error threshold: 100%.
NOTICE:  ‚ùå [B004] ERROR: 4/4 unused index exceed the error threshold: 100%
NOTICE:  ==================================================
NOTICE:  üìä Summary: 3 error(s), 0 warning(s), 0 info
NOTICE:  üî¥ Critical issues found - please review and fix errors
NOTICE:  
NOTICE:  üñ•Ô∏è  CLUSTER CHECKS:
NOTICE:  üîç pglinter found 2 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ùå [C002] ERROR: 6 entries in pg_hba.conf with trust or password authentication method exceed the warning threshold: 6.
NOTICE:  ‚ö†Ô∏è  [C003] WARNING: CLUSTER Encrypted passwords with MD5. 1 : 
password_encryption is md5 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 1 error(s), 1 warning(s), 0 info
NOTICE:  üî¥ Critical issues found - please review and fix errors
NOTICE:  
NOTICE:  üìä TABLE CHECKS:
NOTICE:  üîç pglinter found 6 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [T001] WARNING: TABLE no primary key on table(s) 1 : 
public.users_no_pk 

NOTICE:  ‚ö†Ô∏è  [T002] WARNING: TABLE duplicated index 4 : 
public.products idx idx_composite_1 columns (category,price) is a subset of idx idx_composite_2 columns (category,price)
public.products idx idx_composite_2 columns (category,price) is a subset of idx idx_composite_1 columns (category,price)
public.products idx idx_name_1 columns (name) is a subset of idx idx_name_2 columns (name)
public.products idx idx_name_2 columns (name) is a subset of idx idx_name_1 columns (name) 

NOTICE:  ‚ö†Ô∏è  [T003] WARNING: TABLE unindexed constraint 3 : 
inventory_schema.stock_movements constraint name:stock_movements_stock_id_fkey
public.orders constraint name:fk_product
sales_schema.order_items constraint name:order_items_order_id_fkey 

NOTICE:  ‚ö†Ô∏è  [T004] WARNING: TABLE table with potential missing index 18 : 
audit_schema.user_activity_log:100 % of seq scan
dev_reports.summaries:100 % of seq scan
inventory_schema.product_stock:0 % of seq scan
inventory_schema.stock_movements:100 % of seq scan
prod_analytics.metrics:100 % of seq scan
public.UPPERCASE_ISSUES:100 % of seq scan
public.customer_orders:100 % of seq scan
public.good_table_with_pk:100 % of seq scan
public.inventory_logs:100 % of seq scan
public.items:100 % of seq scan
public.orders:100 % of seq scan
public.product_reviews:100 % of seq scan
public.products:100 % of seq scan
public.user_sessions:100 % of seq scan
public.users_no_pk:100 % of seq scan
sales_schema.order_items:100 % of seq scan
sales_schema.orders:0 % of seq scan
sales_schema.sales_reports:100 % of seq scan 

NOTICE:  ‚ö†Ô∏è  [T008] WARNING: TABLE No role grantee on table. it means that except owner, users will need a direct grant on this table, not through a role. Prefer RBAC access if possible. 8 : 
audit_schema.user_activity_log as no role granted
dev_reports.summaries as no role granted
inventory_schema.product_stock as no role granted
inventory_schema.stock_movements as no role granted
prod_analytics.metrics as no role granted
sales_schema.order_items as no role granted
sales_schema.orders as no role granted
sales_schema.sales_reports as no role granted 

NOTICE:  ‚ö†Ô∏è  [T009] WARNING: TABLE Reserved keywords in object. 2 : 
column:public.items.FROM is a reserved keyword.
column:public.items.SELECT is a reserved keyword. 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 6 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
NOTICE:  
NOTICE:  üóÇÔ∏è  SCHEMA CHECKS:
NOTICE:  üîç pglinter found 2 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [S001] WARNING: SCHEMA No default role grantee on schema {0}.{1}. It means that each time a table is created, you must grant it to roles. 6 : 
audit_schema
dev_reports
inventory_schema
prod_analytics
sales_schema
staging_data 

NOTICE:  ‚ö†Ô∏è  [S002] WARNING: SCHEMA You should not prefix or suffix the schema name with {0}. You may have difficulties when refreshing environments. Prefer prefix or suffix the database name. 3 : 
dev_reports
prod_analytics
staging_data 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 2 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
NOTICE:  
NOTICE:  üéâ All pglinter checks completed successfully!
 check_all 
-----------
 t
(1 row)

-- Test rule management features
SELECT '=== RULE MANAGEMENT ===' AS info;
          info           
-------------------------
 === RULE MANAGEMENT ===
(1 row)

-- Show all rules
SELECT pglinter.show_rules();
NOTICE:  üìã pglinter Rule Status:
NOTICE:  ============================================================
NOTICE:  Code   Status   Name                                    
NOTICE:  ------------------------------------------------------------
NOTICE:  B001   ‚úÖ ON     HowManyTableWithoutPrimaryKey           
NOTICE:  B002   ‚úÖ ON     HowManyRedudantIndex                    
NOTICE:  B003   ‚úÖ ON     HowManyTableWithoutIndexOnFk            
NOTICE:  B004   ‚úÖ ON     HowManyUnusedIndex                      
NOTICE:  B005   ‚úÖ ON     UnsecuredPublicSchema                   
NOTICE:  B006   ‚úÖ ON     HowManyObjectsWithUppercase             
NOTICE:  B007   ‚úÖ ON     HowManyTablesNeverSelected              
NOTICE:  B008   ‚úÖ ON     HowManyTablesWithFkOutsideSchema        
NOTICE:  B015   ‚úÖ ON     HowManyTableSharingSameTrigger          
NOTICE:  C002   ‚úÖ ON     PgHbaEntriesWithMethodTrustOrPasswordShouldNotExists
NOTICE:  C003   ‚úÖ ON     PasswordEncryptionIsMd5                 
NOTICE:  S001   ‚úÖ ON     SchemaWithDefaultRoleNotGranted         
NOTICE:  S002   ‚úÖ ON     SchemaPrefixedOrSuffixedWithEnvt        
NOTICE:  T001   ‚úÖ ON     TableWithoutPrimaryKey                  
NOTICE:  T002   ‚úÖ ON     TableWithRedundantIndex                 
NOTICE:  T003   ‚úÖ ON     TableWithFkNotIndexed                   
NOTICE:  T004   ‚úÖ ON     TableWithPotentialMissingIdx            
NOTICE:  T005   ‚úÖ ON     TableWithFkOutsideSchema                
NOTICE:  T006   ‚úÖ ON     TableWithUnusedIndex                    
NOTICE:  T007   ‚úÖ ON     TableWithFkMismatch                     
NOTICE:  T008   ‚úÖ ON     TableWithRoleNotGranted                 
NOTICE:  T009   ‚úÖ ON     ReservedKeyWord                         
NOTICE:  T010   üî¥ OFF    TableWithSensibleColumn                 
NOTICE:  T015   ‚úÖ ON     TableSharingSameTrigger                 
NOTICE:  ============================================================
NOTICE:  üìä Summary: 23 enabled, 1 disabled
 show_rules 
------------
 t
(1 row)

-- Test some explanations
SELECT pglinter.explain_rule('B001');
NOTICE:  üìñ Rule Explanation for B001
============================================================

üéØ Rule Name: HowManyTableWithoutPrimaryKey
üìã Scope: BASE

üìù Description:
Count number of tables without primary key.

‚ö†Ô∏è  Message Template:
{0}/{1} table(s) without primary key exceed the {2} threshold: {3}%.

üîß How to Fix:
   1. create a primary key or change warning/error threshold
============================================================
 explain_rule 
--------------
 t
(1 row)

SELECT pglinter.explain_rule('T003');
NOTICE:  üìñ Rule Explanation for T003
============================================================

üéØ Rule Name: TableWithFkNotIndexed
üìã Scope: TABLE

üìù Description:
When you delete or update a row in the parent table, the database must check the child table to ensure there are no orphaned records. An index on the foreign key allows for a rapid lookup, ensuring that these checks don't negatively impact performance.

‚ö†Ô∏è  Message Template:
unindexed constraint

üîß How to Fix:
   1. create an index on the child table fk.
============================================================
 explain_rule 
--------------
 t
(1 row)

SELECT pglinter.explain_rule('S002');
NOTICE:  üìñ Rule Explanation for S002
============================================================

üéØ Rule Name: SchemaPrefixedOrSuffixedWithEnvt
üìã Scope: SCHEMA

üìù Description:
The schema is prefixed with one of staging,stg,preprod,prod,sandbox,sbox string. Means that when you refresh your preprod, staging environments from production, you have to rename the target schema from prod_ to stg_ or something like. It is possible, but it is never easy.

‚ö†Ô∏è  Message Template:
You should not prefix or suffix the schema name with {0}. You may have difficulties when refreshing environments. Prefer prefix or suffix the database name.

üîß How to Fix:
   1. Keep the same schema name across environments. Prefer prefix or suffix the database name
============================================================
 explain_rule 
--------------
 t
(1 row)

-- Test output to file functionality
SELECT '=== OUTPUT TO FILE TEST ===' AS info;
            info             
-----------------------------
 === OUTPUT TO FILE TEST ===
(1 row)

SELECT pglinter.perform_base_check('/tmp/integration_base.sarif');
 perform_base_check 
--------------------
 t
(1 row)

SELECT pglinter.perform_table_check('/tmp/integration_table.sarif');
 perform_table_check 
---------------------
 t
(1 row)

SELECT pglinter.perform_schema_check('/tmp/integration_schema.sarif');
 perform_schema_check 
----------------------
 t
(1 row)

-- Test rule disable/enable functionality
SELECT '=== RULE TOGGLE TEST ===' AS info;
           info           
--------------------------
 === RULE TOGGLE TEST ===
(1 row)

-- Disable some rules
SELECT pglinter.disable_rule('B001');
NOTICE:  üî¥ Rule B001 has been disabled
 disable_rule 
--------------
 t
(1 row)

SELECT pglinter.disable_rule('T003');
NOTICE:  üî¥ Rule T003 has been disabled
 disable_rule 
--------------
 t
(1 row)

-- Run checks (should skip disabled rules)
SELECT pglinter.perform_base_check();
NOTICE:  üîç pglinter found 2 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ùå [B003] ERROR: 3/3 table(s) without index on foreign key exceed the error threshold: 100%.
NOTICE:  ‚ùå [B004] ERROR: 4/4 unused index exceed the error threshold: 100%
NOTICE:  ==================================================
NOTICE:  üìä Summary: 2 error(s), 0 warning(s), 0 info
NOTICE:  üî¥ Critical issues found - please review and fix errors
 perform_base_check 
--------------------
 t
(1 row)

SELECT pglinter.perform_table_check();
NOTICE:  üîç pglinter found 5 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [T001] WARNING: TABLE no primary key on table(s) 1 : 
public.users_no_pk 

NOTICE:  ‚ö†Ô∏è  [T002] WARNING: TABLE duplicated index 4 : 
public.products idx idx_composite_1 columns (category,price) is a subset of idx idx_composite_2 columns (category,price)
public.products idx idx_composite_2 columns (category,price) is a subset of idx idx_composite_1 columns (category,price)
public.products idx idx_name_1 columns (name) is a subset of idx idx_name_2 columns (name)
public.products idx idx_name_2 columns (name) is a subset of idx idx_name_1 columns (name) 

NOTICE:  ‚ö†Ô∏è  [T004] WARNING: TABLE table with potential missing index 18 : 
audit_schema.user_activity_log:100 % of seq scan
dev_reports.summaries:100 % of seq scan
inventory_schema.product_stock:0 % of seq scan
inventory_schema.stock_movements:100 % of seq scan
prod_analytics.metrics:100 % of seq scan
public.UPPERCASE_ISSUES:100 % of seq scan
public.customer_orders:100 % of seq scan
public.good_table_with_pk:100 % of seq scan
public.inventory_logs:100 % of seq scan
public.items:100 % of seq scan
public.orders:100 % of seq scan
public.product_reviews:100 % of seq scan
public.products:100 % of seq scan
public.user_sessions:100 % of seq scan
public.users_no_pk:100 % of seq scan
sales_schema.order_items:100 % of seq scan
sales_schema.orders:0 % of seq scan
sales_schema.sales_reports:100 % of seq scan 

NOTICE:  ‚ö†Ô∏è  [T008] WARNING: TABLE No role grantee on table. it means that except owner, users will need a direct grant on this table, not through a role. Prefer RBAC access if possible. 8 : 
audit_schema.user_activity_log as no role granted
dev_reports.summaries as no role granted
inventory_schema.product_stock as no role granted
inventory_schema.stock_movements as no role granted
prod_analytics.metrics as no role granted
sales_schema.order_items as no role granted
sales_schema.orders as no role granted
sales_schema.sales_reports as no role granted 

NOTICE:  ‚ö†Ô∏è  [T009] WARNING: TABLE Reserved keywords in object. 2 : 
column:public.items.FROM is a reserved keyword.
column:public.items.SELECT is a reserved keyword. 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 5 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
 perform_table_check 
---------------------
 t
(1 row)

-- Re-enable rules
SELECT pglinter.enable_rule('B001');
NOTICE:  ‚úÖ Rule B001 has been enabled
 enable_rule 
-------------
 t
(1 row)

SELECT pglinter.enable_rule('T003');
NOTICE:  ‚úÖ Rule T003 has been enabled
 enable_rule 
-------------
 t
(1 row)

-- Final comprehensive check
SELECT pglinter.check_all();
NOTICE:  üîç Running comprehensive pglinter check...
NOTICE:  
NOTICE:  üìã BASE CHECKS:
NOTICE:  üîç pglinter found 3 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ùå [B001] ERROR: 16/18 table(s) without primary key exceed the error threshold: 88%.
NOTICE:  ‚ùå [B003] ERROR: 3/3 table(s) without index on foreign key exceed the error threshold: 100%.
NOTICE:  ‚ùå [B004] ERROR: 4/4 unused index exceed the error threshold: 100%
NOTICE:  ==================================================
NOTICE:  üìä Summary: 3 error(s), 0 warning(s), 0 info
NOTICE:  üî¥ Critical issues found - please review and fix errors
NOTICE:  
NOTICE:  üñ•Ô∏è  CLUSTER CHECKS:
NOTICE:  üîç pglinter found 2 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ùå [C002] ERROR: 6 entries in pg_hba.conf with trust or password authentication method exceed the warning threshold: 6.
NOTICE:  ‚ö†Ô∏è  [C003] WARNING: CLUSTER Encrypted passwords with MD5. 1 : 
password_encryption is md5 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 1 error(s), 1 warning(s), 0 info
NOTICE:  üî¥ Critical issues found - please review and fix errors
NOTICE:  
NOTICE:  üìä TABLE CHECKS:
NOTICE:  üîç pglinter found 6 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [T001] WARNING: TABLE no primary key on table(s) 1 : 
public.users_no_pk 

NOTICE:  ‚ö†Ô∏è  [T002] WARNING: TABLE duplicated index 4 : 
public.products idx idx_composite_1 columns (category,price) is a subset of idx idx_composite_2 columns (category,price)
public.products idx idx_composite_2 columns (category,price) is a subset of idx idx_composite_1 columns (category,price)
public.products idx idx_name_1 columns (name) is a subset of idx idx_name_2 columns (name)
public.products idx idx_name_2 columns (name) is a subset of idx idx_name_1 columns (name) 

NOTICE:  ‚ö†Ô∏è  [T003] WARNING: TABLE unindexed constraint 3 : 
inventory_schema.stock_movements constraint name:stock_movements_stock_id_fkey
public.orders constraint name:fk_product
sales_schema.order_items constraint name:order_items_order_id_fkey 

NOTICE:  ‚ö†Ô∏è  [T004] WARNING: TABLE table with potential missing index 18 : 
audit_schema.user_activity_log:100 % of seq scan
dev_reports.summaries:100 % of seq scan
inventory_schema.product_stock:0 % of seq scan
inventory_schema.stock_movements:100 % of seq scan
prod_analytics.metrics:100 % of seq scan
public.UPPERCASE_ISSUES:100 % of seq scan
public.customer_orders:100 % of seq scan
public.good_table_with_pk:100 % of seq scan
public.inventory_logs:100 % of seq scan
public.items:100 % of seq scan
public.orders:100 % of seq scan
public.product_reviews:100 % of seq scan
public.products:100 % of seq scan
public.user_sessions:100 % of seq scan
public.users_no_pk:100 % of seq scan
sales_schema.order_items:100 % of seq scan
sales_schema.orders:0 % of seq scan
sales_schema.sales_reports:100 % of seq scan 

NOTICE:  ‚ö†Ô∏è  [T008] WARNING: TABLE No role grantee on table. it means that except owner, users will need a direct grant on this table, not through a role. Prefer RBAC access if possible. 8 : 
audit_schema.user_activity_log as no role granted
dev_reports.summaries as no role granted
inventory_schema.product_stock as no role granted
inventory_schema.stock_movements as no role granted
prod_analytics.metrics as no role granted
sales_schema.order_items as no role granted
sales_schema.orders as no role granted
sales_schema.sales_reports as no role granted 

NOTICE:  ‚ö†Ô∏è  [T009] WARNING: TABLE Reserved keywords in object. 2 : 
column:public.items.FROM is a reserved keyword.
column:public.items.SELECT is a reserved keyword. 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 6 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
NOTICE:  
NOTICE:  üóÇÔ∏è  SCHEMA CHECKS:
NOTICE:  üîç pglinter found 2 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [S001] WARNING: SCHEMA No default role grantee on schema {0}.{1}. It means that each time a table is created, you must grant it to roles. 6 : 
audit_schema
dev_reports
inventory_schema
prod_analytics
sales_schema
staging_data 

NOTICE:  ‚ö†Ô∏è  [S002] WARNING: SCHEMA You should not prefix or suffix the schema name with {0}. You may have difficulties when refreshing environments. Prefer prefix or suffix the database name. 3 : 
dev_reports
prod_analytics
staging_data 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 2 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
NOTICE:  
NOTICE:  üéâ All pglinter checks completed successfully!
 check_all 
-----------
 t
(1 row)

-- Clean up
DROP SCHEMA prod_analytics CASCADE;
NOTICE:  drop cascades to table prod_analytics.metrics
DROP SCHEMA dev_reports CASCADE;
NOTICE:  drop cascades to table dev_reports.summaries
DROP SCHEMA staging_data CASCADE;
DROP TABLE items CASCADE;
DROP TABLE "UPPERCASE_ISSUES" CASCADE;
DROP TABLE orders CASCADE;
DROP TABLE products CASCADE;
DROP TABLE users_no_pk CASCADE;
DROP EXTENSION pglinter CASCADE;
